### PR TITLE
Separate auto holds from holds

### PR DESCRIPTION
**Pull request by samplackett**
https://eaflood.atlassian.net/browse/DOC-744

Code to migrate the existing holds to auto holds - 

INSERT INTO public."autoHolds" (frn, "autoHoldCategoryId", added, closed, "migrationId")
SELECT h.frn, h."holdCategoryId", h.added, h.closed, h."migrationId"
FROM holds h
JOIN public."autoHoldCategories" ahc ON h."holdCategoryId" = ahc."autoHoldCategoryId";

UPDATE public."autoHolds" AS ah
SET "marketingYear" = pr."marketingYear"
FROM (
    SELECT *
    FROM public."autoHolds"
) AS ah_update
JOIN (
    SELECT * 
    FROM public."paymentRequests"
    ORDER BY "paymentRequestId" ASC
) AS pr ON ah_update.frn = pr.frn
JOIN public."autoHoldCategories" AS ahc ON ah_update."autoHoldCategoryId" = ahc."autoHoldCategoryId"
WHERE ah.frn = pr.frn
AND ahc."schemeId" = pr."schemeId"
AND NOT EXISTS (
    SELECT 1
    FROM public."completedPaymentRequests" AS cpr
    WHERE pr."paymentRequestId" = cpr."paymentRequestId"
);

DELETE FROM public."holds"
WHERE "holdCategoryId" IN (SELECT "autoHoldCategoryId" FROM public."autoHoldCategories");

### CODE REVIEW

File: app/processing/process-payment-request.js
Comment (johnwatson484) [inline app/processing/process-payment-request.js:20]: Concern with this approach is we might be blocking processing entirely.

Service processes 500 at a time with all non-processable excluded during the original SQL statement.

If the next 500 all have auto hold, then all will be filtered out.  Then they'll be picked up again on next attempt and you've got an endless loop of never processing anything

File: changelog/db.changelog-1.66.xml
Comment (johnwatson484) [inline changelog/db.changelog-1.66.xml:116]: Does this need a FK reference to schemes?

**Review by johnwatson484 — COMMENTED** [review]
Changes look good.  Just a few thoughts.

Firstly, performance testing as discussed is key.  That SQL query is the most intensive process in the service.  Once merged we need to run several performance tests in dev.

For the update script, will it safely get the correct marketing year?  What if there are two blocked with different years, will it get the right one?  If we sort by `paymentRequestId` will it help as presumably the first received caused the hold?

Other thing we need to consider is how we can release this safely.

Is the plan to:
1. merge
2. deploy to dev
3. performance test
4. deploy to test
5. confirm UAT issue resolved
6. run db pipeline to prod
7. run helm pipeline to prod
8. run update script

?

If so, seems ok.  Don't think there'll be any negative impact in each interim deployment step so you've clearly thought it through!  Nice :)

File: changelog/db.changelog-1.66.xml
Comment (samplackett) [reply changelog/db.changelog-1.66.xml:116]: autoHolds doesn't but autoHoldCategories certainly does.

**Comment by samplackett** [comment]
> For the update script, will it safely get the correct marketing year? What if there are two blocked with different years, will it get the right one? If we sort by `paymentRequestId` will it help as presumably the first received caused the hold?

Agreed, we should sort by paymentRequestId ascending - added this to the script. If there are then other payments that should be on hold following the change, they should get re-processed and a new autoHold raised in the table against the same FRN for the new year.

> Other thing we need to consider is how we can release this safely.
> 
> Is the plan to:
> 
> 1. merge
> 2. deploy to dev
> 3. performance test
> 4. deploy to test
> 5. confirm UAT issue resolved
> 6. run db pipeline to prod
> 7. run helm pipeline to prod
> 8. run update script

That's very much the plan in my head 😄

File: app/processing/scheduled/get-pending-payment-requests.js
Comment (johnwatson484) [inline app/processing/scheduled/get-pending-payment-requests.js:7]: Could simplify to `if (!scheduledPaymentRequests.length) {`

File: app/processing/scheduled/get-pending-payment-requests.js
Comment (johnwatson484) [inline app/processing/scheduled/get-pending-payment-requests.js:35]: Current query is `>` not `>=`

File: app/processing/scheduled/get-pending-payment-requests.js
Comment (johnwatson484) [inline app/processing/scheduled/get-pending-payment-requests.js:38]: We shouldn't skip locked row otherwise we'll essentially ignoring them and therefore would defeat the point of this query.

File: app/processing/scheduled/get-pending-payment-requests.js
Comment (johnwatson484) [inline app/processing/scheduled/get-pending-payment-requests.js:22]: Could we use `.map` to simplify this?

**Review by johnwatson484 — COMMENTED** [review]
Added a couple of thoughts for consideration.

Mostly around refining the pending query.  All the hold stuff looks fine

File: app/processing/scheduled/get-pending-payment-requests.js
Comment (samplackett) [reply app/processing/scheduled/get-pending-payment-requests.js]: Similar sort of code but have changed to map which agree is narrowly cleaner

**Review by johnwatson484 — APPROVED** [review]

**Pull request merged**
